File: sdks/ios/Sources/LegacyBaaSSDK/Services/CobolTranspilerService.swift
import Foundation
import Combine

/// COBOL Transpiler Service for iOS SDK
/// Handles all COBOL transpilation operations
@available(iOS 13.0, *)
public class CobolTranspilerService: ObservableObject {
    
    // MARK: - Types
    
    public enum TargetLanguage: String, CaseIterable {
        case solidity = "solidity"
        case corda = "corda"
        case algorand = "algorand"
    }
    
    public enum BankingSystem: String, CaseIterable {
        case fisSystematics = "fis-systematics"
        case fiservDNA = "fiserv-dna"
        case tcsBancs = "tcs-bancs"
        case temenosTransact = "temenos-transact"
    }
    
    public enum BlockchainNetwork: String, CaseIterable {
        case ethereum = "ethereum"
        case corda = "corda"
        case algorand = "algorand"
        case xrp = "xrp"
    }
    
    public enum RiskLevel: String, CaseIterable {
        case low = "LOW"
        case medium = "MEDIUM"
        case high = "HIGH"
        case critical = "CRITICAL"
    }
    
    // MARK: - Models
    
    public struct CobolTranspileRequest {
        public let sourceCode: String
        public let targetLanguage: TargetLanguage
        public let bankingSystem: BankingSystem
        public let blockchainNetwork: BlockchainNetwork
        public let options: TranspileOptions?
        public let compliance: ComplianceConfig?
        
        public init(
            sourceCode: String,
            targetLanguage: TargetLanguage,
            bankingSystem: BankingSystem,
            blockchainNetwork: BlockchainNetwork,
            options: TranspileOptions? = nil,
            compliance: ComplianceConfig? = nil
        ) {
            self.sourceCode = sourceCode
            self.targetLanguage = targetLanguage
            self.bankingSystem = bankingSystem
            self.blockchainNetwork = blockchainNetwork
            self.options = options
            self.compliance = compliance
        }
    }
    
    public struct TranspileOptions {
        public let optimizeForGas: Bool
        public let includeComments: Bool
        public let generateInterface: Bool
        public let strictTypeChecking: Bool
        public let generateValidators: Bool
        public let includeAuditTrail: Bool
        public let autoOptimize: Bool
        public let deployImmediately: Bool
        public let monitorPerformance: Bool
        public let asyncProcessing: Bool
        
        public init(
            optimizeForGas: Bool = true,
            includeComments: Bool = true,
            generateInterface: Bool = true,
            strictTypeChecking: Bool = false,
            generateValidators: Bool = false,
            includeAuditTrail: Bool = false,
            autoOptimize: Bool = false,
            deployImmediately: Bool = false,
            monitorPerformance: Bool = false,
            asyncProcessing: Bool = false
        ) {
            self.optimizeForGas = optimizeForGas
            self.includeComments = includeComments
            self.generateInterface = generateInterface
            self.strictTypeChecking = strictTypeChecking
            self.generateValidators = generateValidators
            self.includeAuditTrail = includeAuditTrail
            self.autoOptimize = autoOptimize
            self.deployImmediately = deployImmediately
            self.monitorPerformance = monitorPerformance
            self.asyncProcessing = asyncProcessing
        }
    }
    
    public struct ComplianceConfig {
        public let configId: String?
        public let riskThreshold: RiskLevel
        public let requireApproval: Bool
        public let generateReport: Bool
        
        public init(
            configId: String? = nil,
            riskThreshold: RiskLevel = .medium,
            requireApproval: Bool = false,
            generateReport: Bool = false
        ) {
            self.configId = configId
            self.riskThreshold = riskThreshold
            self.requireApproval = requireApproval
            self.generateReport = generateReport
        }
    }
    
    public struct CobolTranspileResult {
        public let success: Bool
        public let transactionId: String
        public let jobId: String?
        public let contractAddress: String?
        public let transactionHash: String?
        public let sourceCode: String?
        public let gasEstimate: Int?
        public let deploymentCost: Double?
        public let performance: PerformanceMetrics?
        public let compliance: ComplianceResult?
        public let error: String?
    }
    
    public struct PerformanceMetrics {
        public let compilationTime: Double
        public let deploymentTime: Double
        public let gasUsage: Int
        public let executionTime: Double
        public let throughput: Double
    }
    
    public struct ComplianceResult {
        public let riskScore: Int
        public let screeningResults: [String: Any]
        public let auditTrail: [String: Any]
    }
    
    public struct CobolTemplate {
        public let id: String
        public let name: String
        public let description: String
        public let bankingSystem: BankingSystem
        public let category: String
        public let variables: [TemplateVariable]
        public let version: String
        public let created: Date
        public let updated: Date
    }
    
    public struct TemplateVariable {
        public let name: String
        public let type: String
        public let required: Bool
        public let description: String
    }
    
    public struct CobolStatusUpdate {
        public let jobId: String
        public let progress: Int
        public let stage: String
        public let message: String
        public let timestamp: Date
    }
    
    public struct NetworkSelection {
        public let network: BlockchainNetwork
        public let language: TargetLanguage
        public let reasons: [String]
        public let estimatedCost: Double
        public let estimatedTime: TimeInterval
    }
    
    public struct CobolAnalytics {
        public let totalTranspilations: Int
        public let successRate: Double
        public let averageTime: TimeInterval
        public let byBankingSystem: [String: Int]
        public let byNetwork: [String: Int]
        public let errorTypes: [String: Int]
        public let performanceMetrics: PerformanceMetrics
    }
    
    public struct CobolQuota {
        public let current: Int
        public let limit: Int
        public let remaining: Int
        public let resetDate: Date
        public let overage: Int
        public let overageCost: Double
    }
    
    public struct ValidationResult {
        public let valid: Bool
        public let errors: [ValidationError]
        public let warnings: [String]
        public let suggestions: [String]
    }
    
    public struct ValidationError {
        public let line: Int
        public let column: Int
        public let message: String
        public let severity: ErrorSeverity
    }
    
    public enum ErrorSeverity: String {
        case error = "error"
        case warning = "warning"
        case info = "info"
    }
    
    // MARK: - Properties
    
    private let httpClient: HTTPClient
    private let baseURL: URL
    private var cancellables = Set<AnyCancellable>()
    
    @Published public var isLoading = false
    @Published public var lastError: Error?
    
    // MARK: - Initialization
    
    init(httpClient: HTTPClient, baseURL: URL) {
        self.httpClient = httpClient
        self.baseURL = baseURL.appendingPathComponent("banking")
    }
    
    // MARK: - Public Methods
    
    /// Transpile COBOL code to target blockchain language
    public func transpile(_ request: CobolTranspileRequest) -> AnyPublisher<CobolTranspileResult, Error> {
        isLoading = true
        
        let endpoint = baseURL.appendingPathComponent("transpile")
        
        let requestBody: [String: Any] = [
            "sourceCode": request.sourceCode,
            "targetLanguage": request.targetLanguage.rawValue,
            "bankingSystem": request.bankingSystem.rawValue,
            "blockchainNetwork": request.blockchainNetwork.rawValue,
            "options": request.options?.toDictionary() ?? [:],
            "compliance": request.compliance?.toDictionary() ?? [:]
        ]
        
        return httpClient.post(endpoint: endpoint, body: requestBody)
            .map { (response: APIResponse<CobolTranspileResult>) in
                self.isLoading = false
                return response.data
            }
            .catch { error -> AnyPublisher<CobolTranspileResult, Error> in
                self.isLoading = false
                self.lastError = error
                return Fail(error: error).eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
    }
    
    /// Process multiple COBOL files in batch
    public func batchTranspile(files: [BatchFile], bankingSystem: BankingSystem, options: [String: Any]? = nil) -> AnyPublisher<[CobolTranspileResult], Error> {
        isLoading = true
        
        let endpoint = baseURL.appendingPathComponent("transpile/batch")
        
        let requestBody: [String: Any] = [
            "files": files.map { $0.toDictionary() },
            "bankingSystem": bankingSystem.rawValue,
            "options": options ?? [:]
        ]
        
        return httpClient.post(endpoint: endpoint, body: requestBody)
            .map { (response: APIResponse<[CobolTranspileResult]>) in
                self.isLoading = false
                return response.data
            }
            .catch { error -> AnyPublisher<[CobolTranspileResult], Error> in
                self.isLoading = false
                self.lastError = error
                return Fail(error: error).eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
    }
    
    /// Validate COBOL code without transpilation
    public func validate(sourceCode: String, bankingSystem: BankingSystem) -> AnyPublisher<ValidationResult, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/validate")
        
        let requestBody: [String: Any] = [
            "sourceCode": sourceCode,
            "bankingSystem": bankingSystem.rawValue
        ]
        
        return httpClient.post(endpoint: endpoint, body: requestBody)
            .map { (response: APIResponse<ValidationResult>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Get transpilation job status
    public func getStatus(jobId: String) -> AnyPublisher<CobolStatusUpdate, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/status/\(jobId)")
        
        return httpClient.get(endpoint: endpoint)
            .map { (response: APIResponse<CobolStatusUpdate>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Monitor transpilation status with real-time updates
    public func monitorStatus(jobId: String) -> AnyPublisher<CobolStatusUpdate, Error> {
        Timer.publish(every: 2.0, on: .main, in: .common)
            .autoconnect()
            .flatMap { _ in
                self.getStatus(jobId: jobId)
            }
            .takeWhile { status in
                status.progress < 100
            }
            .eraseToAnyPublisher()
    }
    
    /// Get available COBOL templates
    public func getTemplates(bankingSystem: BankingSystem? = nil, category: String? = nil) -> AnyPublisher<[CobolTemplate], Error> {
        var endpoint = baseURL.appendingPathComponent("transpile/templates")
        
        var queryItems: [URLQueryItem] = []
        if let bankingSystem = bankingSystem {
            queryItems.append(URLQueryItem(name: "bankingSystem", value: bankingSystem.rawValue))
        }
        if let category = category {
            queryItems.append(URLQueryItem(name: "category", value: category))
        }
        
        if !queryItems.isEmpty {
            var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false)!
            components.queryItems = queryItems
            endpoint = components.url!
        }
        
        return httpClient.get(endpoint: endpoint)
            .map { (response: APIResponse<[CobolTemplate]>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Get specific template by ID
    public func getTemplate(templateId: String) -> AnyPublisher<CobolTemplate, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/templates/\(templateId)")
        
        return httpClient.get(endpoint: endpoint)
            .map { (response: APIResponse<CobolTemplate>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Transpile using a template
    public func transpileWithTemplate(templateId: String, variables: [String: Any], targetLanguage: TargetLanguage, blockchainNetwork: BlockchainNetwork, options: [String: Any]? = nil) -> AnyPublisher<CobolTranspileResult, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/templates/\(templateId)")
        
        let requestBody: [String: Any] = [
            "variables": variables,
            "targetLanguage": targetLanguage.rawValue,
            "blockchainNetwork": blockchainNetwork.rawValue,
            "options": options ?? [:]
        ]
        
        return httpClient.post(endpoint: endpoint, body: requestBody)
            .map { (response: APIResponse<CobolTranspileResult>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Select optimal blockchain network
    public func selectOptimalNetwork(criteria: NetworkSelectionCriteria) -> AnyPublisher<NetworkSelection, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/network-selection")
        
        return httpClient.post(endpoint: endpoint, body: criteria.toDictionary())
            .map { (response: APIResponse<NetworkSelection>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Get transpilation analytics
    public func getAnalytics(period: String = "24h") -> AnyPublisher<CobolAnalytics, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/analytics")
        
        var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "period", value: period)]
        
        return httpClient.get(endpoint: components.url!)
            .map { (response: APIResponse<CobolAnalytics>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Get quota information
    public func getQuota() -> AnyPublisher<CobolQuota, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/quota")
        
        return httpClient.get(endpoint: endpoint)
            .map { (response: APIResponse<CobolQuota>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
    
    /// Estimate transpilation cost
    public func estimateCost(request: CostEstimateRequest) -> AnyPublisher<CostEstimate, Error> {
        let endpoint = baseURL.appendingPathComponent("transpile/cost-estimate")
        
        return httpClient.post(endpoint: endpoint, body: request.toDictionary())
            .map { (response: APIResponse<CostEstimate>) in
                response.data
            }
            .eraseToAnyPublisher()
    }
}

// MARK: - Supporting Types

public struct BatchFile {
    public let name: String
    public let content: String
    public let targetLanguage: CobolTranspilerService.TargetLanguage
    public let blockchainNetwork: CobolTranspilerService.BlockchainNetwork
    public let options: [String: Any]?
    
    public init(name: String, content: String, targetLanguage: CobolTranspilerService.TargetLanguage, blockchainNetwork: CobolTranspilerService.BlockchainNetwork, options: [String: Any]? = nil) {
        self.name = name
        self.content = content
        self.targetLanguage = targetLanguage
        self.blockchainNetwork = blockchainNetwork
        self.options = options
    }
    
    func toDictionary() -> [String: Any] {
        return [
            "name": name,
            "content": content,
            "targetLanguage": targetLanguage.rawValue,
            "blockchainNetwork": blockchainNetwork.rawValue,
            "options": options ?? [:]
        ]
    }
}

public struct NetworkSelectionCriteria {
    public let contractComplexity: String
    public let transactionVolume: String
    public let latencyRequirement: String
    public let costSensitivity: String
    
    public init(contractComplexity: String, transactionVolume: String, latencyRequirement: String, costSensitivity: String) {
        self.contractComplexity = contractComplexity
        self.transactionVolume = transactionVolume
        self.latencyRequirement = latencyRequirement
        self.costSensitivity = costSensitivity
    }
    
    func toDictionary() -> [String: Any] {
        return [
            "contractComplexity": contractComplexity,
            "transactionVolume": transactionVolume,
            "latencyRequirement": latencyRequirement,
            "costSensitivity": costSensitivity
        ]
    }
}

public struct CostEstimateRequest {
    public let sourceCodeLength: Int
    public let targetLanguage: String
    public let blockchainNetwork: String
    public let complexity: String?
    
    public init(sourceCodeLength: Int, targetLanguage: String, blockchainNetwork: String, complexity: String? = nil) {
        self.sourceCodeLength = sourceCodeLength
        self.targetLanguage = targetLanguage
        self.blockchainNetwork = blockchainNetwork
        self.complexity = complexity
    }
    
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "sourceCodeLength": sourceCodeLength,
            "targetLanguage": targetLanguage,
            "blockchainNetwork": blockchainNetwork
        ]
        if let complexity = complexity {
            dict["complexity"] = complexity
        }
        return dict
    }
}

public struct CostEstimate {
    public let estimatedCost: Double
    public let breakdown: CostBreakdown
    public let currency: String
}

public struct CostBreakdown {
    public let baseCost: Double
    public let complexityMultiplier: Double
    public let networkFee: Double
    public let deploymentCost: Double
}

// MARK: - Extensions

extension CobolTranspilerService.TranspileOptions {
    func toDictionary() -> [String: Any] {
        return [
            "optimizeForGas": optimizeForGas,
            "includeComments": includeComments,
            "generateInterface": generateInterface,
            "strictTypeChecking": strictTypeChecking,
            "generateValidators": generateValidators,
            "includeAuditTrail": includeAuditTrail,
            "autoOptimize": autoOptimize,
            "deployImmediately": deployImmediately,
            "monitorPerformance": monitorPerformance,
            "async": asyncProcessing
        ]
    }
}

extension CobolTranspilerService.ComplianceConfig {
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "riskThreshold": riskThreshold.rawValue,
            "requireApproval": requireApproval,
            "generateReport": generateReport
        ]
        if let configId = configId {
            dict["configId"] = configId
        }
        return dict
    }
}

// MARK: - Codable Extensions

extension CobolTranspilerService.CobolTranspileResult: Codable {
    enum CodingKeys: String, CodingKey {
        case success, transactionId, jobId, contractAddress, transactionHash, sourceCode, gasEstimate, deploymentCost, performance, compliance, error
    }
}

extension CobolTranspilerService.PerformanceMetrics: Codable {}
extension CobolTranspilerService.ComplianceResult: Codable {}
extension CobolTranspilerService.CobolTemplate: Codable {}
extension CobolTranspilerService.TemplateVariable: Codable {}
extension CobolTranspilerService.CobolStatusUpdate: Codable {}
extension CobolTranspilerService.NetworkSelection: Codable {}
extension CobolTranspilerService.CobolAnalytics: Codable {}
extension CobolTranspilerService.CobolQuota: Codable {}
extension CobolTranspilerService.ValidationResult: Codable {}
extension CobolTranspilerService.ValidationError: Codable {}
extension CostEstimate: Codable {}
extension CostBreakdown: Codable {}

extension CobolTranspilerService.TargetLanguage: Codable {}
extension CobolTranspilerService.BankingSystem: Codable {}
extension CobolTranspilerService.BlockchainNetwork: Codable {}
extension CobolTranspilerService.RiskLevel: Codable {}
extension CobolTranspilerService.ErrorSeverity: Codable {}

File: sdks/android/src/main/java/com/legacybaas/android/sdk/services/CobolTranspilerService.kt
package com.legacybaas.android.sdk.services

import com.legacybaas.android.sdk.LegacyBaaSClient
import com.legacybaas.android.sdk.models.cobol.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.delay
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * COBOL Transpiler Service for Android SDK
 * Handles all COBOL transpilation operations with Kotlin coroutines
 */
class CobolTranspilerService internal constructor(
    private val client: LegacyBaaSClient
) {
    
    /**
     * Target programming languages for transpilation
     */
    enum class TargetLanguage(val value: String) {
        SOLIDITY("solidity"),
        CORDA("corda"),
        ALGORAND("algorand")
    }
    
    /**
     * Supported banking systems
     */
    enum class BankingSystem(val value: String) {
        FIS_SYSTEMATICS("fis-systematics"),
        FISERV_DNA("fiserv-dna"),
        TCS_BANCS("tcs-bancs"),
        TEMENOS_TRANSACT("temenos-transact")
    }
    
    /**
     * Supported blockchain networks
     */
    enum class BlockchainNetwork(val value: String) {
        ETHEREUM("ethereum"),
        CORDA("corda"),
        ALGORAND("algorand"),
        XRP("xrp")
    }
    
    /**
     * Risk assessment levels
     */
    enum class RiskLevel(val value: String) {
        LOW("LOW"),
        MEDIUM("MEDIUM"),
        HIGH("HIGH"),
        CRITICAL("CRITICAL")
    }
    
    /**
     * Transpile COBOL code to target blockchain language
     */
    suspend fun transpile(request: CobolTranspileRequest): CobolTranspileResult {
        val requestBody = mapOf(
            "sourceCode" to request.sourceCode,
            "targetLanguage" to request.targetLanguage.value,
            "bankingSystem" to request.bankingSystem.value,
            "blockchainNetwork" to request.blockchainNetwork.value,
            "options" to (request.options?.toMap() ?: emptyMap<String, Any>()),
            "compliance" to (request.compliance?.toMap() ?: emptyMap<String, Any>())
        )
        
        return client.performRequest<CobolTranspileResult>(
            "/banking/transpile",
            LegacyBaaSClient.HttpMethod.POST,
            requestBody
        )
    }
    
    /**
     * Process multiple COBOL files in batch
     */
    suspend fun batchTranspile(
        files: List<BatchFile>,
        bankingSystem: BankingSystem,
        options: Map<String, Any>? = null
    ): List<CobolTranspileResult> {
        val requestBody = mapOf(
            "files" to files.map { it.toMap() },
            "bankingSystem" to bankingSystem.value,
            "options" to (options ?: emptyMap())
        )
        
        return client.performRequest<List<CobolTranspileResult>>(
            "/banking/transpile/batch",
            LegacyBaaSClient.HttpMethod.POST,
            requestBody
        )
    }
    
    /**
     * Validate COBOL code without transpilation
     */
    suspend fun validate(
        sourceCode: String,
        bankingSystem: BankingSystem
    ): ValidationResult {
        val requestBody = mapOf(
            "sourceCode" to sourceCode,
            "bankingSystem" to bankingSystem.value
        )
        
        return client.performRequest<ValidationResult>(
            "/banking/transpile/validate",
            LegacyBaaSClient.HttpMethod.POST,
            requestBody
        )
    }
    
    /**
     * Get transpilation job status
     */
    suspend fun getStatus(jobId: String): CobolStatusUpdate {
        return client.performRequest<CobolStatusUpdate>(
            "/banking/transpile/status/$jobId",
            LegacyBaaSClient.HttpMethod.GET
        )
    }
    
    /**
     * Monitor transpilation status with real-time updates
     */
    fun monitorStatus(jobId: String): Flow<CobolStatusUpdate> = flow {
        while (true) {
            val status = getStatus(jobId)
            emit(status)
            
            if (status.progress >= 100) {
                break
            }
            
            delay(2000) // Poll every 2 seconds
        }
    }
    
    /**
     * Get available COBOL templates
     */
    suspend fun getTemplates(
        bankingSystem: BankingSystem? = null,
        category: String? = null
    ): List<CobolTemplate> {
        val queryParams = mutableListOf<String>()
        bankingSystem?.let { queryParams.add("bankingSystem=${it.value}") }
        category?.let { queryParams.add("category=$it") }
        
        val endpoint = "/banking/transpile/templates" + 
            if (queryParams.isNotEmpty()) "?${queryParams.joinToString("&")}" else ""
        
        return client.performRequest<List<CobolTemplate>>(
            endpoint,
            LegacyBaaSClient.HttpMethod.GET
        )
    }
    
    /**
     * Get specific template by ID
     */
    suspend fun getTemplate(templateId: String): CobolTemplate {
        return client.performRequest<CobolTemplate>(
            "/banking/transpile/templates/$templateId",
            LegacyBaaSClient.HttpMethod.GET
        )
    }
    
    /**
     * Transpile using a template
     */
    suspend fun transpileWithTemplate(
        templateId: String,
        variables: Map<String, Any>,
        targetLanguage: TargetLanguage,
        blockchainNetwork: BlockchainNetwork,
        options: Map<String, Any>? = null
    ): CobolTranspileResult {
        val requestBody = mapOf(
            "variables" to variables,
            "targetLanguage" to targetLanguage.value,
            "blockchainNetwork" to blockchainNetwork.value,
            "options" to (options ?: emptyMap())
        )
        
        return client.performRequest<CobolTranspileResult>(
            "/banking/transpile/templates/$templateId",
            LegacyBaaSClient.HttpMethod.POST,
            requestBody
        )
    }
    
    /**
     * Select optimal blockchain network
     */
    suspend fun selectOptimalNetwork(
        criteria: NetworkSelectionCriteria
    ): NetworkSelection {
        return client.performRequest<NetworkSelection>(
            "/banking/transpile/network-selection",
            LegacyBaaSClient.HttpMethod.POST,
            criteria.toMap()
        )
    }
    
    /**
     * Get transpilation analytics
     */
    suspend fun getAnalytics(period: String = "24h"): CobolAnalytics {
        return client.performRequest<CobolAnalytics>(
            "/banking/transpile/analytics?period=$period",
            LegacyBaaSClient.HttpMethod.GET
        )
    }
    
    /**
     * Get quota information
     */
    suspend fun getQuota(): CobolQuota {
        return client.performRequest<CobolQuota>(
            "/banking/transpile/quota",
            LegacyBaaSClient.HttpMethod.GET
        )
    }
    
    /**
     * Estimate transpilation cost
     */
    suspend fun estimateCost(request: CostEstimateRequest): CostEstimate {
        return client.performRequest<CostEstimate>(
            "/banking/transpile/cost-estimate",
            LegacyBaaSClient.HttpMethod.POST,
            request.toMap()
        )
    }
}

// Data classes for COBOL transpiler operations

@Serializable
data class CobolTranspileRequest(
    val sourceCode: String,
    val targetLanguage: CobolTranspilerService.TargetLanguage,
    val bankingSystem: CobolTranspilerService.BankingSystem,
    val blockchainNetwork: CobolTranspilerService.BlockchainNetwork,
    val options: TranspileOptions? = null,
    val compliance: ComplianceConfig? = null
)

@Serializable
data class TranspileOptions(
    val optimizeForGas: Boolean = true,
    val includeComments: Boolean = true,
    val generateInterface: Boolean = true,
    val strictTypeChecking: Boolean = false,
    val generateValidators: Boolean = false,
    val includeAuditTrail: Boolean = false,
    val autoOptimize: Boolean = false,
    val deployImmediately: Boolean = false,
    val monitorPerformance: Boolean = false,
    val asyncProcessing: Boolean = false
) {
    fun toMap(): Map<String, Any> = mapOf(
        "optimizeForGas" to optimizeForGas,
        "includeComments" to includeComments,
        "generateInterface" to generateInterface,
        "strictTypeChecking" to strictTypeChecking,
        "generateValidators" to generateValidators,
        "includeAuditTrail" to includeAuditTrail,
        "autoOptimize" to autoOptimize,
        "deployImmediately" to deployImmediately,
        "monitorPerformance" to monitorPerformance,
        "async" to asyncProcessing
    )
}

@Serializable
data class ComplianceConfig(
    val configId: String? = null,
    val riskThreshold: CobolTranspilerService.RiskLevel = CobolTranspilerService.RiskLevel.MEDIUM,
    val requireApproval: Boolean = false,
    val generateReport: Boolean = false
) {
    fun toMap(): Map<String, Any> = buildMap {
        put("riskThreshold", riskThreshold.value)
        put("requireApproval", requireApproval)
        put("generateReport", generateReport)
        configId?.let { put("configId", it) }
    }
}

@Serializable
data class CobolTranspileResult(
    val success: Boolean,
    val transactionId: String,
    val jobId: String? = null,
    val contractAddress: String? = null,
    val transactionHash: String? = null,
    val sourceCode: String? = null,
    val gasEstimate: Int? = null,
    val deploymentCost: Double? = null,
    val performance: PerformanceMetrics? = null,
    val compliance: ComplianceResult? = null,
    val error: String? = null
)

@Serializable
data class PerformanceMetrics(
    val compilationTime: Double,
    val deploymentTime: Double,
    val gasUsage: Int,
    val executionTime: Double,
    val throughput: Double
)

@Serializable
data class ComplianceResult(
    val riskScore: Int,
    val screeningResults: JsonObject,
    val auditTrail: JsonObject
)

@Serializable
data class CobolTemplate(
    val id: String,
    val name: String,
    val description: String,
    val bankingSystem: CobolTranspilerService.BankingSystem,
    val category: String,
    val variables: List<TemplateVariable>,
    val version: String,
    val created: String,
    val updated: String
)

@Serializable
data class TemplateVariable(
    val name: String,
    val type: String,
    val required: Boolean,
    val description: String
)

@Serializable
data class CobolStatusUpdate(
    val jobId: String,
    val progress: Int,
    val stage: String,
    val message: String,
    val timestamp: String
)

@Serializable
data class NetworkSelection(
    val network: CobolTranspilerService.BlockchainNetwork,
    val language: CobolTranspilerService.TargetLanguage,
    val reasons: List<String>,
    val estimatedCost: Double,
    val estimatedTime: Double
)

@Serializable
data class CobolAnalytics(
    val totalTranspilations: Int,
    val successRate: Double,
    val averageTime: Double,
    val byBankingSystem: Map<String, Int>,
    val byNetwork: Map<String, Int>,
    val errorTypes: Map<String, Int>,
    val performanceMetrics: PerformanceMetrics
)

@Serializable
data class CobolQuota(
    val current: Int,
    val limit: Int,
    val remaining: Int,
    val resetDate: String,
    val overage: Int,
    val overageCost: Double
)

@Serializable
data class ValidationResult(
    val valid: Boolean,
    val errors: List<ValidationError>,
    val warnings: List<String>,
    val suggestions: List<String>
)

@Serializable
data class ValidationError(
    val line: Int,
    val column: Int,
    val message: String,
    val severity: ErrorSeverity
)

@Serializable
enum class ErrorSeverity(val value: String) {
    ERROR("error"),
    WARNING("warning"),
    INFO("info")
}

@Serializable
data class BatchFile(
    val name: String,
    val content: String,
    val targetLanguage: CobolTranspilerService.TargetLanguage,
    val blockchainNetwork: CobolTranspilerService.BlockchainNetwork,
    val options: Map<String, Any>? = null
) {
    fun toMap(): Map<String, Any> = mapOf(
        "name" to name,
        "content" to content,
        "targetLanguage" to targetLanguage.value,
        "blockchainNetwork" to blockchainNetwork.value,
        "options" to (options ?: emptyMap())
    )
}

@Serializable
data class NetworkSelectionCriteria(
    val contractComplexity: String,
    val transactionVolume: String,
    val latencyRequirement: String,
    val costSensitivity: String
) {
    fun toMap(): Map<String, Any> = mapOf(
        "contractComplexity" to contractComplexity,
        "transactionVolume" to transactionVolume,
        "latencyRequirement" to latencyRequirement,
        "costSensitivity" to costSensitivity
    )
}

@Serializable
data class CostEstimateRequest(
    val sourceCodeLength: Int,
    val targetLanguage: String,
    val blockchainNetwork: String,
    val complexity: String? = null
) {
    fun toMap(): Map<String, Any> = buildMap {
        put("sourceCodeLength", sourceCodeLength)
        put("targetLanguage", targetLanguage)
        put("blockchainNetwork", blockchainNetwork)
        complexity?.let { put("complexity", it) }
    }
}

@Serializable
data class CostEstimate(
    val estimatedCost: Double,
    val breakdown: CostBreakdown,
    val currency: String
)

@Serializable
data class CostBreakdown(
    val baseCost: Double,
    val complexityMultiplier: Double,
    val networkFee: Double,
    val deploymentCost: Double
)